covid19_sfbayarea/data/contra_costa.py

```diff
@@ -0,0 +1,415 @@
+        day = api.parse_date(x['qNum'])
+        tests[day]['cumul_tests'] = y['qNum']
+
+    # There's no chart with total positive tests, so we use the *rate* to
+    # calculate it. Note also that this is a rolling average, not the actual
+    # rate on the day, so this is not the most accurate. :(
+    #
+    # This chart is a "stacked" chart (it has two series -- rate and equity
+    # metric rate), so is structured in a more complex way.
+    # The data is a list of dates, where each has a `qSubNodes` list with one
+    # item from each series. Those items each have their own `qSubNodes` list
+    # with one item that contains the actual value.
+    positive_rate_chart = api.get_data(CHART_IDS['test_positivity'])
+    positive_rate_data = positive_rate_chart['qLayout']['qHyperCube']['qStackedDataPages'][0]['qData'][0]['qSubNodes']
+    for node in positive_rate_data:
+        day = api.parse_date(node['qValue'])
+        # In the first list of subnodes:
+        #   0 = "% Positive Equity Metric"
+        #   1 = "% Tested Positive 7-Day Average"
+        # In the nested list of subnodes, there's only one item, which has the
+        # actual value.
+        rate = node['qSubNodes'][1]['qSubNodes'][0]['qValue']
+        # Store the rate; we need to do anothe pass to calculate daily tests
+        # before we can use the rate to calculate daily positives.
+        tests[day]['positivity'] = rate
+
+    # Clean up, sanity-check, and put the results in order.
+    result = []
+    total = 0
+    total_positive = 0
+    found_first = False
+    for day in sorted(tests.keys()):
+        record = tests[day]
+        # Skip over all items earlier than the first day with total tests.
+        # (It might be possible for the positivity rate chart to start first.)
+        if not found_first:
+            if record['cumul_tests'] > -1:
+                found_first = True
+                total = record['cumul_tests']
+            continue
+
+        record['date'] = day.isoformat()
+        record['tests'] = record['cumul_tests'] - total
+        total = record['cumul_tests']
+
+        # Estimate positive tests from positivity rate.
+        if record['positivity'] > -1:
+            record['positive'] = round(record['tests'] * record['positivity'])
+            total_positive += record['positive']
+        else:
+            record['positive'] = 0
+        record['cumul_pos'] = total_positive
+
+        result.append(record)
+
+    return result
+
+
+def get_case_totals(api: QlikClient) -> Dict:
+    # Gender and age groups are published as counts, but race and ethnicity are
+    # published as percentages, so we need thet total to calculate counts.
+    gender = get_cases_by_gender(api)
+    total = sum(gender.values())
+
+    return {
+        'gender': gender,
+        'age_group': get_cases_by_age(api),
+        'race_eth': get_cases_by_race(api, total),
+        'ethnicity': get_cases_by_ethnicity(api, total),
```

Comment:
I think the way you've handled this is sensible, given how the data is presented. It would be nice to know how the categories in the "Ethnicity" and "Race" charts are associated, but my guess is that the county just treats these as separate fields.

Reply:
Yeah, I’m pretty sure these are totally unrelated, as far as the county is concerned. 😞